### PR TITLE
Min-Max heap performance improvement

### PR DESCRIPTION
**Pull request by GuyAv46**
simplifying and improving internal computations of the min-max heap, for better readability and performance

benchmark comparing changes to master code [here](https://snapshots.raintank.io/dashboard/snapshot/7lwS8N9Ep5D47IaEk4QlH0SPy4fRgjlI)

### CODE REVIEW

**Comment by filipecosta90** [comment]
### Automated performance analysis summary

This comment was automatically generated given there is performance data available.

In summary:
- Detected a total of 1 highly unstable benchmarks.

You can check a comparison in detail via the [grafana link](https://benchmarksrediscom.grafana.net/d//3Ejv2wZnk/?var-branch=master&var-branch=guyav-minmax_heap_perf)

### Comparison between master and guyav-minmax_heap_perf.

Time Period from 30 days ago. (environment used: oss-standalone)

|      Test Case      |Baseline master (median obs. +- std.dev)|Comparison guyav-minmax_heap_perf (median obs. +- std.dev)|% change (higher-better)|            Note             |
|---------------------|----------------------------------------|----------------------------------------------------------|------------------------|-----------------------------|
|search-numeric-sortby| 3789 +- 40.6% UNSTABLE (7 datapoints)  | 85 +- nan%  (1 datapoints)                               |-97.8%                  |UNSTABLE (very high variance)|

File: src/util/minmax_heap.c [hunk 1]
```diff
@@ -14,7 +14,9 @@
 
 #include "rmalloc.h"
 
-#define is_min(n) ((log2_32(n) & 1) == 0)
+// parity of the number of leading zeros, AKA parity of log2(n)
```
Comment (ephraimfeldblum) [inline src/util/minmax_heap.c:17]: Good. But for future proofing, please add another couple words explaining why this is correct.

File: src/util/minmax_heap.c [hunk 2]
```diff
@@ -24,17 +26,11 @@
 
 #define heap_max(h, x, y) (heap_gt(h, x, y) ? h->data[x] : h->data[y])
 
-static const int tab32[32] = {0, 9,  1,  10, 13, 21, 2,  29, 11, 14, 16, 18, 22, 25, 3, 30,
-                              8, 12, 20, 28, 15, 17, 24, 7,  19, 27, 23, 6,  26, 5,  4, 31};
+#define choose_from_3(fn, a, b, c) \
```
Comment (ephraimfeldblum) [inline src/util/minmax_heap.c:29]: These macros should be functions. These aren't generically typed.

File: src/util/minmax_heap.c [hunk 3]
```diff
@@ -84,23 +80,47 @@ static void bubbleup(heap_t* h, int i) {
...
-  return min_idx;
+int index_max_child_grandchild(heap_t* h, int i) {
+  int a = first_child(i);
+  int b = second_child(i);
+  int c = first_child(a);
+  int d = second_child(a);
+  int e = first_child(b);
+  int f = second_child(b);
+
+  switch (highest_index(h, i)) {
```
Comment (ephraimfeldblum) [inline src/util/minmax_heap.c:108]: Code duplication.

File: src/util/minmax_heap.c [hunk 4]
```diff
@@ -182,7 +201,7 @@ void* mmh_pop_min(heap_t* h) {
   if (h->count > 1) {
     void* d = h->data[1];
     h->data[1] = h->data[h->count--];
-    trickledown(h, 1);
+    trickledown_min(h, 1);
```
Comment (ephraimfeldblum) [inline src/util/minmax_heap.c:204]: Good spotting this.

File: src/util/minmax_heap.c (hunk 3 above)
Comment (GuyAv46) [reply src/util/minmax_heap.c:108]: the calculation of all the children and grandchildren is done once, as you can see [here](https://godbolt.org/z/75c818ae7)

**Review by DvirDukhan — APPROVED** [review]

File: src/util/minmax_heap.c [hunk 5]
```diff
@@ -14,35 +14,37 @@
...
+ * A node is a min node if it's level (depth) is odd (and the root has a depth 1).
+ * With our array representation, a node is a min node if the log2 floor of its index is even.
+ * (log2 floor of 1 is 0 - min, log2 floor of 2 is 1 - max, log2 floor of 3 is 1 - max, log2 floor of 4 is 2 - min, etc.)
+ * A quick way to calculate the log2 floor of a number is to count the leading zeros in its binary representation:
+ * for a 32 bit number, the log2 floor is "31 - the number of leading zeros". `__builtin_clz` does exactly that (clz = count leading zeros).
+ * since we only care about the parity of the log2 floor, we can just check the LSB of the number of leading zeros:
+ * n is a min node <=> log2(n) % 2 == 0 <=> (31 - __builtin_clz(n)) % 2 == 0 <=> __builtin_clz(n) % 2 == 1
+ * So we can simply check for:
+ */
+#define is_min(n) (__builtin_clz(n) & 1)
```
[the remark this answers is not in the document]
Comment (ephraimfeldblum) [reply src/util/minmax_heap.c:32]: This was addressed in the comment above. log(0) is mathematically undefined, so it's no less well defined than the former. The heap is 1-indexed so all arguments to the function must be >0 regardless.

**Review by ephraimfeldblum — APPROVED** [review]
LGTM

**Pull request merged**